### PR TITLE
Transformation

### PR DESCRIPTION
**Pull request by swetha-aggidevara**

### CODE REVIEW

File: generics/services/questionset.js [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+          })
+        } else if (data.statusCode == 200) {
+          let readRes = JSON.parse(data.body)
+          return resolve(readRes)
+        }
+      }
+      request.get(url, { headers: headers }, readQuestionSetCallBack)
+
+    } catch (error) {
+      console.log("errt", error)
```
Comment (aks30) [inline generics/services/questionset.js:67]: @swetha-aggidevara please remove console logs.

File: generics/services/questionset.js [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    } catch (error) {
+      return reject(error);
+    }
+  })
+}
+
+module.exports = {
+  copyQuestionSet: copyQuestionSet,
+  readQuestionSet: readQuestionSet,
+  updateQuestionSetHierarchy: updateQuestionSetHierarchy,
```
Comment (aks30) [inline generics/services/questionset.js:172]: @swetha-aggidevara for what use case we need updateQuestionSetHierarchy ?

File: templates/criteriaAndAssessment.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
+
+const baseAssessment = {
```
Comment (aks30) [inline templates/criteriaAndAssessment.js:2]: @swetha-aggidevara we should move this inside module of criteria and do add comments on each of the object types.

File: templates/questionTemp.js [hunk 4]
```diff
@@ -0,0 +1,154 @@
+
+
+
+const baseQuestion = {
```
Comment (aks30) [inline templates/questionTemp.js:4]: @swetha-aggidevara we should move this inside module of questions and do add comments on each of the object types.

File: module/surveys/helper.js [hunk 5]
```diff
@@ -1808,4 +1809,401 @@ module.exports = class SurveysHelper {
         })
     }
 
+    static getDetailsByLink2(link= "", userId= "", token= "", roleInformation= {},version = "") {
```
Comment (aks30) [inline module/surveys/helper.js:1812]: @swetha-aggidevara are we not using existing functions?

File: module/surveys/helper.js [hunk 6]
```diff
@@ -1808,4 +1809,401 @@ module.exports = class SurveysHelper {
...
+                    success: false,
+                    message: error.message,
+                    data: false
+                });
+            }
+        });
+    }
+
+
+    static detailsV4(surveyId = "", userId= "", submissionId = "", roleInformation = {}, userToken ="") {
```
Comment (aks30) [inline module/surveys/helper.js:1940]: @swetha-aggidevara are we not using existing functions?

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on Aug 1

File: module/surveys/helper.js (hunk 5 above)
Comment (swetha-aggidevara) [reply module/surveys/helper.js:1812]: Existing function is being used by other api's and this function gets the data from API

File: module/surveys/helper.js (hunk 6 above)
Comment (swetha-aggidevara) [reply module/surveys/helper.js:1940]: Existing function is being used by other api's and this function gets the data from API

File: generics/services/questionset.js (hunk 2 above)
Comment (swetha-aggidevara) [reply generics/services/questionset.js:172]: This is Aishwarya's code usage, we are using the same file

File: package.json [hunk 7]
```diff
@@ -59,6 +61,8 @@
     "morgan": "^1.9.0",
     "p-each-series": "^2.1.0",
     "path": "^0.12.7",
+    "querystring": "^0.2.1",
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply package.json:64]: This is being used in the migration script

File: controllers/v1/observationSubmissionsController.js [hunk 8]
```diff
@@ -99,7 +102,17 @@ module.exports = class ObservationSubmissions extends Abstract {
     return new Promise(async (resolve, reject) => {
 
       try {
-        let observationDocument = await observationsHelper.observationDocuments({
+          const cacheData = await cache
+            .get(`submissionCreate:${req.params._id}:${req.params.entityId}`)
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply controllers/v1/observationSubmissionsController.js:106]: What if there a request with the same solutionId but different entityId ?

File: controllers/v1/observationSubmissionsController.js [hunk 9]
```diff
@@ -168,6 +184,18 @@ module.exports = class ObservationSubmissions extends Abstract {
 
         solutionDocument = solutionDocument[0];
 
+            const migratedId = solutionDocument?.migratedId;
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply controllers/v1/observationSubmissionsController.js:187]: Should it be update in migration script also migratedId ===> referenceQuestionSetId ?

File: controllers/v1/observationSubmissionsController.js [hunk 10]
```diff
@@ -168,6 +172,18 @@ module.exports = class ObservationSubmissions extends Abstract {
 
         solutionDocument = solutionDocument[0];
 
+            const referenceQuestionSetId = solutionDocument?.referenceQuestionSetId;
+
+            if (!referenceQuestionSetId) {
```
Comment (aks30) [inline controllers/v1/observationSubmissionsController.js:177]: @swetha-aggidevara Please do proper indentation

File: controllers/v1/observationSubmissionsController.js [hunk 11]
```diff
@@ -331,14 +318,14 @@ module.exports = class ObservationSubmissions extends Abstract {
...
+  
       return resolve({
           message: responseMessage,
           result: observations
       });
-
       } catch (error) {
         return reject({
-          status: error.status || httpStatusCode.internal_server_error.status,
+          status: error?.status || httpStatusCode.internal_server_error.status,
```
Comment (aks30) [inline controllers/v1/observationSubmissionsController.js:328]: @swetha-aggidevara not sure if the syntax is correct do check

File: controllers/v1/solutionsController.js [hunk 12]
```diff
@@ -2285,92 +2298,29 @@ module.exports = class Solutions extends Abstract {
...
-                        });
-                    }
-                }
-            });
-        });
+  
+            submissionDocumentEvidences = solutionDocument.evidenceMethods;
+  
+            let evidences = {};
+            if (!!referenceQuestionSetId) {
```
Comment (aks30) [inline controllers/v1/solutionsController.js:2317]: @swetha-aggidevara do recheck this condition.

File: controllers/v1/surveysController.js [hunk 13]
```diff
@@ -659,15 +656,15 @@ module.exports = class Surveys extends Abstract {
                     req.params._id,
                     req.userDetails.userId,
                     req.userDetails.userToken,
-                    bodyData
+                    bodyData,
+                    "transformation"
```
Comment (aks30) [inline controllers/v1/surveysController.js:660]: @swetha-aggidevara what does this key do ?

File: generics/services/knowledge-platform.js [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+const CREATION_PORTAL_URL = process.env.CREATION_PORTAL_URL
+
+
+const copyQuestionSet = function (copyReq, questionSetId) {
+  const copyQuestionSetUrl = CREATION_PORTAL_URL + messageConstants.endpoints.COPY_QUESTION_SET + "/" + questionSetId;
+  const headers = {
+    "content-type": "application/json",
+    "Authorization": process.env.CREATION_PORTAL_AUTHORIZATION_KEY
+  }
+  const options = {
```
Comment (aks30) [inline generics/services/knowledge-platform.js:12]: @swetha-aggidevara just make this a common constant variable in the file

File: generics/services/knowledge-platform.js [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+const publishQuestionSet = function (questionsetId) {
+  return new Promise((resolve, reject) => {
+    try {
+
+      let publishUrl = `${CREATION_PORTAL_URL}${messageConstants.endpoints.PUBLISH_QUESTION_SET}/${questionsetId}`;
+      const headers = {
+        "content-type": "application/json",
+        "Authorization": process.env.CREATION_PORTAL_AUTHORIZATION_KEY
+      }
+      async function publishQuestionSetCallBack(err, data) {
```
Comment (aks30) [inline generics/services/knowledge-platform.js:112]: @swetha-aggidevara if we can make this as consistent as other functions.

File: module/surveys/helper.js [hunk 16]
```diff
@@ -787,7 +788,8 @@ module.exports = class SurveysHelper {
                     userId,
                     validateSurvey.data.submissionId,
                     roleInformation,
-                    token
+                    token,
+                    type
```
Comment (aks30) [inline module/surveys/helper.js:792]: @swetha-aggidevara we have added this type everywhere not sure what is the use case if you can explain, please.

File: module/surveys/helper.js [hunk 17]
```diff
@@ -859,13 +861,27 @@ module.exports = class SurveysHelper {
...
+                    ],   
                 )
 
                 if (!solutionDocument.length) {
                     throw new Error(messageConstants.apiResponses.SOLUTION_NOT_FOUND)
                 }
 
+                let referenceQuestionSetId = solutionDocument.referenceQuestionSetId;
+
+                if (!!type) {
```
Comment (aks30) [inline module/surveys/helper.js:877]: @swetha-aggidevara do check this condition please.

File: module/questions/transformationHelper.js [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+      assessmentTemplate.assessment.evidences[0].sections[0].questions =
+        evidences?.evidenceSections || [];
+      assessmentTemplate.assessment.evidences[0].sections[0].code = "SQ";
+      assessmentTemplate.assessment.evidences[0].code =
+        assessmentTemplate.assessment.evidences[0].externalId = "SF";
+      assessmentTemplate.assessment.evidences[0].description =
+        questionSetHierarchy?.descripton || "";
+
+      await cache.setEx(
+        `${solutionDocument._id}:${referenceQuestionSetId}`,
```
Comment (aks30) [inline module/questions/transformationHelper.js:60]: @swetha-aggidevara just solutionId is fine right, why do we need questionSetId as well ?

File: module/questions/transformationHelper.js [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+const { baseAssessment, assessmentTemplate } = require(MODULES_BASE_PATH +
+  "/assessments/assessmentTemplate");
+
+module.exports = class Transformation {
+  static getQuestionSetHierarchy(
+    submissionDocumentCriterias,
+    solutionDocument,
+    isPageQuestionsRequired = true
+  ) {
+    return new Promise(async (resolve, reject) => {
```
Comment (aks30) [inline module/questions/transformationHelper.js:18]: @swetha-aggidevara Please use try catch and return statements. Refer - https://github.com/shikshalokam/ml-survey-service/blob/master/module/observations/helper.js

File: module/questions/transformationHelper.js [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+              str = start.slice(endIndex);
+            } else str = "";
+          }
+
+          childQuestion[key] = questionData;
+        } else if (key === "validation") {
+          const obj = {};
+          for (const cKey in keyData) {
+            if (cKey === "required") {
+              const requir = childData?.interactions?.validation?.required;
```
Comment (ankitpws) [inline module/questions/transformationHelper.js:269]: @swetha-aggidevara  too many if else if please update with switch case

File: module/questions/transformationHelper.js [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+        }
+      }
+      resolve(childQuestion);
+    });
+  }
+
+  static getMatrixQuestions(criteria) {
+    const matrixObj = {};
+    for (let key in questionType.matrix) {
+      if (key === "instanceIdentifier") {
```
Comment (ankitpws) [inline module/questions/transformationHelper.js:344]: @swetha-aggidevara update if else if to switch case

**Review by ankitpws — CHANGES_REQUESTED** [review]
Reviewed on 9 -May by Ankit

File: generics/services/knowledge-platform.js [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+      }
+      request.post(copyQuestionSetUrl, options, copyQuestionSetCallback)
+
+    } catch (error) {
+      return reject(error);
+    }
+  })
+
+}
+
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/knowledge-platform.js:41]: @swetha-aggidevara please add comments and function documentation to all functions in this file

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on 9-may-2023

File: controllers/v1/solutionsController.js [hunk 23]
```diff
@@ -86,8 +86,10 @@ module.exports = class Solutions extends Abstract {
         criteriaDocument.forEach(eachCriteria => {
           let levelsDescription = {};
 
-          for (let k in eachCriteria.rubric.levels) {
-            levelsDescription[k] = eachCriteria.rubric.levels[k].description;
+          if (eachCriteria?.rubric?.levels) {
```
[the remark this answers is not in the document]
Comment (swetha-aggidevara) [reply controllers/v1/solutionsController.js:89]: It's an existing code.

**Pull request merged**